sched/irq/irq_attach_thread.c

```diff
@@ -0,0 +1,214 @@
+/****************************************************************************
+ * sched/irq/irq_attach_thread.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <errno.h>
+#include <stdio.h>
+
+#include <nuttx/irq.h>
+#include <nuttx/kthread.h>
+
+#include "irq/irq.h"
+#include "sched/sched.h"
+
+/****************************************************************************
+ * Privte Types
+ ****************************************************************************/
+
+/* This is the type of the list of interrupt handlers, one for each IRQ.
+ * This type provided all of the information necessary to irq_dispatch to
+ * transfer control to interrupt handlers after the occurrence of an
+ * interrupt.
+ */
+
+struct irq_thread_info_s
+{
+  xcpt_t handler;     /* Address of the interrupt handler */
+  xcpt_t isrthread;   /* Address of the interrupt thread handler */
+  FAR void *arg;      /* The argument provided to the interrupt handler. */
+  sem_t sem;          /* irq sem used to notify irq thread */
+  pid_t threadid;     /* irq threadid */
+};
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct irq_thread_info_s g_irq_thread_vector[NR_IRQS];
```

Comment:
```suggestion
static struct irq_thread_info_s g_irq_thread_vector[NR_IRQS];
```

move the global vector to stack to save the data size

Reply:
I didn't get the meaning of this line of comment. Could you explain it further?

Reply:
```
static int irq_default_handler(int irq, FAR void *regs, FAR void *arg)
{
	struct irq_thread_info_s *irq_info = arg;
}


static int isr_thread_main(int argc, FAR char *argv[])
{
	struct irq_thread_info_s irq_info;
	irq_info.xxx =
	irq_info.yyy =
	irq_info.zzz = 
	irq_attach(irq, irq_default_handler, &irq_info);
}
```

Reply:
> ```
> static int irq_default_handler(int irq, FAR void *regs, FAR void *arg)
> {
> 	struct irq_thread_info_s *irq_info = arg;
> }
> 
> 
> static int isr_thread_main(int argc, FAR char *argv[])
> {
> 	struct irq_thread_info_s irq_info;
> 	irq_info.xxx =
> 	irq_info.yyy =
> 	irq_info.zzz = 
> 	irq_attach(irq, irq_default_handler, &irq_info);
> }
> ```

I tried the modification you recommended, but found an issue.
the timing of the irq_attach call is relatively early, during the system boot phase, where the scheduler is locked and
it's impossible to immediately start an isr thread.

For example,
A common sequence of calls is:
irq_attach
up_enable_irq

We must ensure that irq_attach is synchronous and completes the binding.
Otherwise, an exception will be reported when the interrupt is responded to.

This is because if the interrupt is enabled (up_enable_irq) before the handler is fully bound (irq_attach completes),
the system may not be prepared to handle the interrupt properly, leading to undefined behavior or crashes.

Reply:
> the timing of the irq_attach call is relatively early, during the system boot phase, where the scheduler is locked and
it's impossible to immediately start an isr thread.

1. NuttX will start the scheduler before entering the idle state
https://github.com/apache/nuttx/blob/master/sched/init/nx_start.c#L762
2. The system startup does not depends on any interrupts, such as `timer`. Applications(snh) could be started even if the system has no interrupts
> We must ensure that irq_attach is synchronous and completes the binding.
> Otherwise, an exception will be reported when the interrupt is responded to.
> This is because if the interrupt is enabled (up_enable_irq) before the handler is fully bound (irq_attach completes),
the system may not be prepared to handle the interrupt properly, leading to undefined behavior or crashes.

3. You should check your driver design, Why does the system crash after delayed binding?

Reply:
The current nuttx design initializes time interrupts before the scheduling unlock. This issue arises if the time interrupts use isrthread.

As for the g_irq_thread_vector datasize, I have another optimization patch that will configure the maximum number of global arrays based on the actually used number of interrupts, rather than relying on irq numbers.

Reply:
> The current nuttx design initializes time interrupts before the scheduling unlock. This issue arises if the time interrupts use isrthread.
So you enabled irq before irqthread started? Shouldn't irq be enabled after the thread is started?

> As for the g_irq_thread_vector datasize, I have another optimization patch that will configure the maximum number of global arrays based on the actually used number of interrupts, rather than relying on irq numbers.

If your implementation still depends on `static arrays` or `dynamic heap`s (mallloc?), then you don't need to submit it, the stack alloc will be more elegant.

Reply:
> I tried the modification you recommended, but found an issue.

It also looks like the argument parameter is just saved as a pointer to the irq_info instance.  That will be a stale pointer at the time of the interrupt

Reply:
ok I update the patch